### PR TITLE
TEngine installation rules fix for static build

### PR DESCRIPTION
**Pull request by liqi-c**
```
Xbuild_shared:ARMv7=OFF
Xbuild_shared:ARMv8=OFF
force_builders=docs,ARMv7,ARMv8,Android pack
```

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by liqi-c** [comment]
PR for modify issue #17042

**Comment by h6197627** [comment]
Hi, @liqi-c 
Thanks, your PR fixes CMake error and installs libtengine.a in a proper directory. However it still additionally installs include directory and lib directory into root install directory. I found that it can be fixed by changing line
`add_subdirectory("${OCV_TENGINE_SOURCE_PATH}" "${OCV_TENGINE_DIR}/build")`
to
`add_subdirectory("${OCV_TENGINE_SOURCE_PATH}" "${OCV_TENGINE_DIR}/build" EXCLUDE_FROM_ALL)`

File: 3rdparty/libtengine/tengine.cmake [hunk 1]
```diff
@@ -88,4 +86,8 @@ if(BUILD_TENGINE)
 		message(WARNING "TENGINE: Missing 'CMakeLists.txt' in source code package: ${OCV_TENGINE_SOURCE_PATH}")
 		SET(HAVE_TENGINE 1)
```
Comment (h6197627) [inline 3rdparty/libtengine/tengine.cmake:87]: I think it is already done on line 63

Comment (liqi-c) [reply 3rdparty/libtengine/tengine.cmake:87]: @h6197627 You look very carefully, thank you , i will modify this .

**Comment by asmorkalov** [comment]
The solution does not work for me. The branch build for ARM Linux with cross-compiler fails on CMake stage:
```
$ ~/Projects/opencv-build-arm64$ cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/linux/aarch64-gnu.toolchain.cmake -DGCC_COMPILER_VERSION=7 ../opencv
...
CMake Error at cmake/OpenCVUtils.cmake:1185 (install):
  install TARGETS given target "tengine" which does not exist in this
  directory.
Call Stack (most recent call first):
  3rdparty/libtengine/tengine.cmake:89 (ocv_install_target)
  cmake/OpenCVFindTengine.cmake:59 (include)
  CMakeLists.txt:691 (include)


-- Found Tengine include: /home/alexander/Projects/opencv-build-arm64/3rdparty/libtengine/Tengine-2f3cd86217f3530c8e4a82f3ed5af14c7a4e3943/core/include
-- Found Tengine libraries: tengine
```

**Comment by liqi-c** [comment]
> The solution does not work for me. The branch build for ARM Linux with cross-compiler fails on CMake stage:
> 
> ```
> $ ~/Projects/opencv-build-arm64$ cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/linux/aarch64-gnu.toolchain.cmake -DGCC_COMPILER_VERSION=7 ../opencv
> ...
> CMake Error at cmake/OpenCVUtils.cmake:1185 (install):
>   install TARGETS given target "tengine" which does not exist in this
>   directory.
> Call Stack (most recent call first):
>   3rdparty/libtengine/tengine.cmake:89 (ocv_install_target)
>   cmake/OpenCVFindTengine.cmake:59 (include)
>   CMakeLists.txt:691 (include)
> 
> 
> -- Found Tengine include: /home/alexander/Projects/opencv-build-arm64/3rdparty/libtengine/Tengine-2f3cd86217f3530c8e4a82f3ed5af14c7a4e3943/core/include
> -- Found Tengine libraries: tengine
> ```

I'm sorry , I didn't repeat your problem . can you provide me a complete log ?   and
 cmake verision ?
below is my log file . thanks 
[log_20200609.txt](https://github.com/opencv/opencv/files/4750275/log_20200609.txt)

**Comment by asmorkalov** [comment]
Environment: Ubuntu 18.04 with `crossbuild-essential-arm64` package installed.
cmake: 3.10.2.
Code version: 4.3.0-346-g3dab884186

CMake command:
```
 cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/linux/aarch64-gnu.toolchain.cmake -DGCC_COMPILER_VERSION=7 ../opencv
```
CMake log: [cmake.log](https://github.com/opencv/opencv/files/4750888/cmake.log)

**Comment by asmorkalov** [comment]
Android configuration fails too.
CMake command line:
```
cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk-r18b/build/cmake/android.toolchain.cmake -DANDROID_NATIVE_API_LEVEL=21 -DANDROID_TOOLCHAIN=clang -DANDROID_ABI=armeabi-v7a -DANDROID_ARM_NEON=ON ../opencv
```
Log is attached: 
[cmake.log](https://github.com/opencv/opencv/files/4751832/cmake.log)

**Comment by liqi-c** [comment]
> Android configuration fails too.
> CMake command line:
> 
> ```
> cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk-r18b/build/cmake/android.toolchain.cmake -DANDROID_NATIVE_API_LEVEL=21 -DANDROID_TOOLCHAIN=clang -DANDROID_ABI=armeabi-v7a -DANDROID_ARM_NEON=ON ../opencv
> ```
> 
> Log is attached:
> [cmake.log](https://github.com/opencv/opencv/files/4751832/cmake.log)

Can you help try the new modify by me . This modify maybe can adapt any environment ?

**Comment by h6197627** [comment]
> Android configuration fails too.
> CMake command line:
> 
> ```
> cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk-r18b/build/cmake/android.toolchain.cmake -DANDROID_NATIVE_API_LEVEL=21 -DANDROID_TOOLCHAIN=clang -DANDROID_ABI=armeabi-v7a -DANDROID_ARM_NEON=ON ../opencv
> ```
> 
> Log is attached:
> [cmake.log](https://github.com/opencv/opencv/files/4751832/cmake.log)

Hm, I just rechecked previous commit (https://github.com/opencv/opencv/pull/17468/commits/3dab884186d59336bac387b2b0665db88c7d9600) and got everything fine by running:
`
cmake -DBUILD_SHARED_LIBS=OFF -DWITH_TENGINE=ON -DCMAKE_TOOLCHAIN_FILE=/home/user/Android/sdk/ndk-bundle/build/cmake/android.toolchain.cmake -DANDROID_SDK=/home/user/Android/sdk -DANDROID_NATIVE_API_LEVEL=21 -DANDROID_TOOLCHAIN=clang -DANDROID_ABI=armeabi-v7a -DANDROID_ARM_NEON=ON ..
`
```
-- TENGINE:--  Build Tengine from source code. 
-- TENGINE: Download: 2f3cd86217f3530c8e4a82f3ed5af14c7a4e3943.zip
-- TENGINE DOWNLOAD success . 
-- CMAKE_INSTALL_PREFIX = /home/user/Downloads/opencv-4.3.0/build/install
-- CONFIG_ARCH_ARM32 = ON
-- The ASM-ATT compiler identification is GNU
-- Found assembler: /home/user/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-as
-- Found Tengine include: /home/user/Downloads/opencv-4.3.0/build/3rdparty/libtengine/Tengine-2f3cd86217f3530c8e4a82f3ed5af14c7a4e3943/core/include
-- Found Tengine libraries: tengine
```

Maybe it is something related with CMake? Mine version is 3.16.3

**Comment by h6197627** [comment]
Well, indeed, it is CMake related. I removed my manually build CMake 3.16.3 and installed version 3.10.2 that is default for Ubuntu 18.04 and reproduced @asmorkalov error.

File: 3rdparty/libtengine/tengine.cmake [hunk 2]
```diff
@@ -69,23 +69,21 @@ if(BUILD_TENGINE)
...
 			   SET(CONFIG_ARCH_ARM32 ON)
 		elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL aarch64) ## AARCH64
 			   SET(CONFIG_ARCH_ARM64 ON)
 		endif()
-		SET(Tengine_LIB "tengine" CACHE INTERNAL "")
 	endif()
 
 	SET(BUILT_IN_OPENCV ON) ## set for tengine compile discern .
 	SET(Tengine_INCLUDE_DIR  "${OCV_TENGINE_SOURCE_PATH}/core/include" CACHE INTERNAL "")
+	SET(Tengine_LIB "${CMAKE_BINARY_DIR}/lib/${ANDROID_ABI}/libtengine.a" CACHE INTERNAL "")
```
Comment (asmorkalov) [inline 3rdparty/libtengine/tengine.cmake:83]: The solution is not good enough too. There are 2 issues:
- TEngine library is not installed. It means that it's missing from Android SDK and other builds that are done with cross-compiler.
- There is build path leak to final CMake config:
```
~/Projects/OpenCV/opencv-master-arm64/install/lib/cmake/opencv4$ grep -Rin tengine
OpenCVModules.cmake:146:  INTERFACE_LINK_LIBRARIES "\$<LINK_ONLY:opencv_core>;\$<LINK_ONLY:opencv_imgproc>;opencv_core;opencv_imgproc;\$<LINK_ONLY:dl>;\$<LINK_ONLY:m>;\$<LINK_ONLY:pthread>;\$<LINK_ONLY:rt>;\$<LINK_ONLY:tegra_hal>;\$<LINK_ONLY:libprotobuf>;/home/alexander/Projects/OpenCV/opencv-master-arm64/lib//libtengine.a"
```

**Comment by asmorkalov** [comment]
I recommend you to look at other 3rd party libraries inside OpenCV repo like JPEG. They work properly for both dynamic and static build, including cross-compilation.

**Comment by h6197627** [comment]
I tried new commit: it configures without errors, however it does not install libtengine.a for static build while running
> make install

**Comment by liqi-c** [comment]
> I tried new commit: it configures without errors, however it does not install libtengine.a for static build while running
> 
> > make install

 when build static lib ,tengine will be include in libopencv_dnn.so , maybe this is not a problem ?

**Comment by h6197627** [comment]
Hi, @liqi-c 
when we build static version with  -DBUILD_SHARED_LIBS=OFF we get libopencv_dnn.a which additionally requires libtengine.a while linking into user projects.
When we build shared OpenCV version with -DBUILD_SHARED_LIBS=ON libtengine.a shouldn't be installed as it is already included into libopencv_dnn.so as you wrote

**Comment by liqi-c** [comment]
del&nbsp; EXCLUDE_FROM_ALL&nbsp; when add_subdirectory can solve this problem .&nbsp;&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "h6197627"<notifications@github.com&gt;; 
发送时间: 2020年6月15日(星期一) 下午4:44
收件人: "opencv/opencv"<opencv@noreply.github.com&gt;; 
抄送: "琦"<liqi-c@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [opencv/opencv] TEngine installation rules fix for static build (#17468)





 
Hi, @liqi-c
 when we build static version with  -DBUILD_SHARED_LIBS=OFF we get libopencv_dnn.a which additionally requires libtengine.a while linking into user projects.
 When we build shared OpenCV version with -DBUILD_SHARED_LIBS=ON libtengine.a shouldn't be installed as it is already included into libopencv_dnn.so as you wrote

**Comment by h6197627** [comment]
That way it installs into root installation directory instead of **sdk/native/3rdparty/libs/{archtecture}**. Also it installs include directory which is not needed.
And it installs for both static/shared configurations (for shared it shouldn't)

**Comment by alalek** [comment]
`EXCLUDE_FROM_ALL` brokes `install()` rules.

Also 3rdparty libs should not know anything about OpenCV's install layout - they should use own variables for install subdirs.
This variable can be specified by OpenCV integration scripts:
```
set(TENGINE_INSTALL_ARGS TARGET tengine EXPORT OpenCVModules ARCHIVE DESTINATION "${OPENCV_3P_LIB_INSTALL_PATH}" COMPONENT dev)
add_subdirectory(...)  # without EXCLUDE_FROM_ALL

... in tengine ...

if(DEFINED TENGINE_INSTALL_ARGS)  # custom override
    install(${TENGINE_INSTALL_ARGS})
else()
    ... own rules ...
endif()
```

Or add variable to skip install rules in tengine part and wrote own in OpenCV scripts.

**Comment by h6197627** [comment]
Works for me

**Comment by liqi-c** [comment]
@h6197627 @asmorkalov @alalek  thanks very much .

**Pull request merged**